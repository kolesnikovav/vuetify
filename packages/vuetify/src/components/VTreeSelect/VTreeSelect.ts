import VSelect from '../VSelect/VSelect'

export default VSelect.extend({
    name: 'v-tree-select'

})
