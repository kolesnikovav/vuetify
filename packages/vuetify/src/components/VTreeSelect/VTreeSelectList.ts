// Utils
import mixins from '../../util/mixins'
import { provide as RegistrableProvide } from '../../mixins/registrable'

// Mixins
import Themeable from '../../mixins/themeable'
import Colorable from '../../mixins/colorable'

export default mixins(
    RegistrableProvide('treeselectlist'),
    Themeable, Colorable
    /* @vue/component */
  ).extend({
    name: 'v-tree-select-list',

    provide (): object {
      return { treeselectlist: this }
    }
})
