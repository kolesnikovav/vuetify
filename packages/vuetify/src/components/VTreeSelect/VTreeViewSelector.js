import VTreeView from '../VTreeview/VTreeview'

export default VTreeView.extend({
  name: 'v-tree-view-selector',
  props: {
    selectedItems: {
      type: Array,
      default: []
    }
  },
  data: () => ({
    selectedCache: new Set()
  }),
  watch: {
    selectedItems () {
      if (this.selectedItems.length < 1) {
        this.clearSelection()
      }
    }
  },
  methods: {
    clearSelection () {
      this.selectedCache.clear()
      Object.keys(this.nodes).forEach(key => {
        this.nodes[key].isSelected = false
        this.nodes[key].isIndeterminate = false
      })
    }
  }
})
