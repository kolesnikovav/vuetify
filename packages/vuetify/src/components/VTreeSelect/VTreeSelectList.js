/* uses instead of VSelectList from base component */
// Utils
import mixins from '../../util/mixins'
import { VListTile, VListTileContent, VListTileTitle } from '../VList'

// Mixins
import Themeable from '../../mixins/themeable'
import Colorable from '../../mixins/colorable'
import VTreeViewSelector from './VTreeViewSelector'
import { VTreeviewNodeProps } from '../VTreeview/VTreeviewNode'

export default mixins(
  Themeable, Colorable
  /* @vue/component */
).extend({
  name: 'v-tree-select-list',
  props: {
    selectedItems: [],
    noDataText: String,
    dense: Boolean,
    multiple: Boolean,
    items: {
      type: Array,
      default: () => ([])
    },
    openAll: Boolean,
    returnObject: {
      type: Boolean,
      default: false // TODO: Should be true in next major
    },
    value: {
      type: Array,
      default: () => ([])
    },
    search: String,
    filter: {
      type: Function,
      default: undefined
    },
    ...VTreeviewNodeProps
  },
  computed: {
    staticNoDataTile () {
      const tile = {
        on: {
          mousedown: e => e.preventDefault() // Prevent onBlur from being called
        }
      }
      return this.$createElement(VListTile, tile, [
        this.genTileNoDataContent()
      ])
    }
  },
  methods: {
    genTileNoDataContent () {
      const innerHTML = this.noDataText
      return this.$createElement(VListTileContent,
        [this.$createElement(VListTileTitle, {
          domProps: { innerHTML }
        })]
      )
    }
  },
  render (h) {
    const children = []
    if (!this.items || !Array.isArray(this.items) || this.items.length < 1) {
      children.push(this.$slots['no-data'] || this.staticNoDataTile)
    }
    this.$slots['prepend-item'] && children.unshift(this.$slots['prepend-item'])
    const childrenAppend = []
    this.$slots['append-item'] && childrenAppend.push(this.$slots['append-item'])

    return this.$createElement('div', {
      staticClass: 'v-select-list v-card',
      'class': this.themeClasses
    }, [
      children,
      this.$createElement(VTreeViewSelector, {
        props: {
          dense: this.dense,
          items: this.items,
          selectable: true,
          returnObject: true,
          selectOnly: true,
          selectedItems: this.selectedItems,
          openAll: this.openAll
        },
        on: {
          input: e => {
            this.$emit('select', e)
          }
        }
      }), childrenAppend
    ])
  }
})
