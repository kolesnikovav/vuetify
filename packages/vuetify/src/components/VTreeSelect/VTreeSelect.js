import VAutocomplete from '../VAutocomplete/VAutocomplete'
import VSelect from '../VSelect/VSelect'
import VTreeSelectList from './VTreeSelectList'
import { VTreeviewNodeProps } from '../VTreeview/VTreeviewNode'
import { consoleError } from '../../util/console'

export default VAutocomplete.extend({
  name: 'v-tree-select',
  props: {
    autocomplete: {
      type: Boolean,
      default: false
    },
    openAll: Boolean,
    ...VSelect.options.props,
    ...VAutocomplete.options.props,
    ...VTreeviewNodeProps
  },
  computed: {
    listData () {
      const data = VSelect.options.computed.listData.call(this)
      Object.assign(data.props, { selectedItems: this.selectedItems,
        openAll: this.openAll,
        ...VTreeviewNodeProps })
      Object.assign(data.on, {
        select: e => {
          this.selectItems(e)
        }
      })
      return data
    },
    staticList () {
      if (this.$slots['no-data'] || this.$slots['prepend-item'] || this.$slots['append-item']) {
        consoleError('assert: staticList should not be called if slots are used')
      }
      return this.$createElement(VTreeSelectList, this.listData)
    }
  },
  methods: {
    register () {},
    genListWithSlot () {
      const slots = ['prepend-item', 'no-data', 'append-item']
        .filter(slotName => this.$slots[slotName])
        .map(slotName => this.$createElement('template', {
          slot: slotName
        }, this.$slots[slotName]))
      // Requires destructuring due to Vue
      // modifying the `on` property when passed
      // as a referenced object
      return this.$createElement(VTreeSelectList, {
        ...this.listData
      }, slots)
    },
    selectItems (items) {
      this.selectedItems = items
      if (!this.multiple) {
        this.isMenuActive = false
      }
    },
    clearableCallback () {
      this.internalValue = null
      this.$refs.input.internalValue = ''
      this.searchInput = ''
      this.$refs.input.value = ''
      this.$nextTick(() => this.$refs.input.focus())
    }
  }
})
