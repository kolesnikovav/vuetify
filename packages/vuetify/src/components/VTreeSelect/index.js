import VTreeSelect from './VTreeSelect'
import VTreeSelectList from './VTreeSelectList'

export { VTreeSelect, VTreeSelectList }

export default {
  $_vuetify_subcomponents: {
    VTreeSelect,
    VTreeSelectList
  }
}
