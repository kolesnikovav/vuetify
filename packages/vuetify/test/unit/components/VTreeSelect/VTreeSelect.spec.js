import { test } from '@/test'
import { keyCodes } from '@/util/helpers'
import VTreeSelect from '@/components/VTreeSelect'

test('VTreeSelect.js', ({ mount, compileToFunctions }) => {
  const app = document.createElement('div')
  app.setAttribute('data-app', true)
  document.body.appendChild(app)

  const staticTree = [
      {id: 1, name: 'foo', children: [
          {id: 2, name: 'bar'},
          {id: 3, name: 'baz'},
          {id: 4, name: 'wiz'}
      ]},
      {id:5 , name: 'Alex'},
      {id:6 , name: 'Petr'}
  ]

  it('should render correctly and match snapshot when toggle menu', async () => {
    const wrapper = mount(VTreeSelect, {
      propsData: {
        items: staticTree
      }
    })
    const icon = wrapper.first('.v-icon')
    const slot = wrapper.first('.v-input__slot')

    expect(wrapper.vm.isMenuActive).toBe(false)

    slot.trigger('click')
    expect(wrapper.vm.isMenuActive).toBe(true)

    slot.trigger('click')
    expect(wrapper.vm.isMenuActive).toBe(true)

    const treeview = wrapper.first('.v-treeview')
    expect(treeview.html()).toMatchSnapshot()


  })

})