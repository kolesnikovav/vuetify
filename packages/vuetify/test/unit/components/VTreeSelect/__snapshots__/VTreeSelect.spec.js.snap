// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VTreeSelect.js should render correctly and match snapshot when toggle menu 1`] = `

<div class="v-treeview theme--light">
  <div class="v-treeview-node">
    <div class="v-treeview-node__root">
      <i aria-hidden="true"
         class="v-icon v-treeview-node__toggle v-icon--link material-icons theme--light"
      >
        arrow_drop_down
      </i>
      <i aria-hidden="true"
         class="v-icon v-treeview-node__checkbox v-icon--link material-icons theme--light"
      >
        check_box_outline_blank
      </i>
      <div class="v-treeview-node__content">
        <div class="v-treeview-node__label">
          foo
        </div>
      </div>
    </div>
  </div>
  <div class="v-treeview-node v-treeview-node--leaf">
    <div class="v-treeview-node__root">
      <i aria-hidden="true"
         class="v-icon v-treeview-node__checkbox v-icon--link material-icons theme--light"
      >
        check_box_outline_blank
      </i>
      <div class="v-treeview-node__content">
        <div class="v-treeview-node__label">
          Alex
        </div>
      </div>
    </div>
  </div>
  <div class="v-treeview-node v-treeview-node--leaf">
    <div class="v-treeview-node__root">
      <i aria-hidden="true"
         class="v-icon v-treeview-node__checkbox v-icon--link material-icons theme--light"
      >
        check_box_outline_blank
      </i>
      <div class="v-treeview-node__content">
        <div class="v-treeview-node__label">
          Petr
        </div>
      </div>
    </div>
  </div>
</div>

`;

exports[`VTreeSelect.js should render correctly and match snapshot with ItemText props custom 1`] = `

<div class="v-treeview theme--light">
  <div class="v-treeview-node">
    <div class="v-treeview-node__root">
      <i aria-hidden="true"
         class="v-icon v-treeview-node__toggle v-icon--link material-icons theme--light"
      >
        arrow_drop_down
      </i>
      <i aria-hidden="true"
         class="v-icon v-treeview-node__checkbox v-icon--link material-icons theme--light"
      >
        check_box_outline_blank
      </i>
      <div class="v-treeview-node__content">
        <div class="v-treeview-node__label">
        </div>
      </div>
    </div>
  </div>
  <div class="v-treeview-node v-treeview-node--leaf">
    <div class="v-treeview-node__root">
      <i aria-hidden="true"
         class="v-icon v-treeview-node__checkbox v-icon--link material-icons theme--light"
      >
        check_box_outline_blank
      </i>
      <div class="v-treeview-node__content">
        <div class="v-treeview-node__label">
        </div>
      </div>
    </div>
  </div>
  <div class="v-treeview-node v-treeview-node--leaf">
    <div class="v-treeview-node__root">
      <i aria-hidden="true"
         class="v-icon v-treeview-node__checkbox v-icon--link material-icons theme--light"
      >
        check_box_outline_blank
      </i>
      <div class="v-treeview-node__content">
        <div class="v-treeview-node__label">
        </div>
      </div>
    </div>
  </div>
</div>

`;

exports[`VTreeSelect.js should render correctly and match snapshot with openAll items 1`] = `

<div class="v-treeview theme--light">
  <div class="v-treeview-node">
    <div class="v-treeview-node__root">
      <i aria-hidden="true"
         class="v-icon v-treeview-node__toggle v-icon--link material-icons theme--light v-treeview-node__toggle--open"
      >
        arrow_drop_down
      </i>
      <i aria-hidden="true"
         class="v-icon v-treeview-node__checkbox v-icon--link material-icons theme--light"
      >
        check_box_outline_blank
      </i>
      <div class="v-treeview-node__content">
        <div class="v-treeview-node__label">
          foo
        </div>
      </div>
    </div>
    <div class="v-treeview-node__children">
      <div class="v-treeview-node v-treeview-node--leaf">
        <div class="v-treeview-node__root">
          <i aria-hidden="true"
             class="v-icon v-treeview-node__checkbox v-icon--link material-icons theme--light"
          >
            check_box_outline_blank
          </i>
          <div class="v-treeview-node__content">
            <div class="v-treeview-node__label">
              bar
            </div>
          </div>
        </div>
      </div>
      <div class="v-treeview-node v-treeview-node--leaf">
        <div class="v-treeview-node__root">
          <i aria-hidden="true"
             class="v-icon v-treeview-node__checkbox v-icon--link material-icons theme--light"
          >
            check_box_outline_blank
          </i>
          <div class="v-treeview-node__content">
            <div class="v-treeview-node__label">
              baz
            </div>
          </div>
        </div>
      </div>
      <div class="v-treeview-node v-treeview-node--leaf">
        <div class="v-treeview-node__root">
          <i aria-hidden="true"
             class="v-icon v-treeview-node__checkbox v-icon--link material-icons theme--light"
          >
            check_box_outline_blank
          </i>
          <div class="v-treeview-node__content">
            <div class="v-treeview-node__label">
              wiz
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div class="v-treeview-node v-treeview-node--leaf">
    <div class="v-treeview-node__root">
      <i aria-hidden="true"
         class="v-icon v-treeview-node__checkbox v-icon--link material-icons theme--light"
      >
        check_box_outline_blank
      </i>
      <div class="v-treeview-node__content">
        <div class="v-treeview-node__label">
          Alex
        </div>
      </div>
    </div>
  </div>
  <div class="v-treeview-node v-treeview-node--leaf">
    <div class="v-treeview-node__root">
      <i aria-hidden="true"
         class="v-icon v-treeview-node__checkbox v-icon--link material-icons theme--light"
      >
        check_box_outline_blank
      </i>
      <div class="v-treeview-node__content">
        <div class="v-treeview-node__label">
          Petr
        </div>
      </div>
    </div>
  </div>
</div>

`;
